### PR TITLE
[spec_v2] Enable trtllm_mha draft-extend CUDA graph with v2 semantics

### PR DESCRIPTION
**Pull request by YAMY1234**
## Motivation

`DRAFT_EXTEND_V2` currently cannot use the draft-extend CUDA graph path with `trtllm_mha` because the v2 Eagle draft worker whitelist does not include `TRTLLMHAAttnBackend`.

The previous V2 graph replay computed full-logits `softmax + topk/reduce` work, while `EagleDraftWorker` later recomputed `softmax + fast_topk` on the selected logits it actually consumes. The graph-produced `ret.topk_p` and `ret.topk_index` are therefore dead work in the current V2 worker contract.

This PR enables draft-extend CUDA graph for `TRTLLMHAAttnBackend`, fixes the associated V2 replay metadata, and removes that graph-side dead top-k work without changing TRTLLM MHA attention kernel routing.

## Modifications

- Add `TRTLLMHAAttnBackend` to the v2 Eagle draft worker CUDA graph whitelist.
- Include `DRAFT_EXTEND_V2` in the `trtllm_mha` draft-extend CUDA graph capture/replay metadata branches.
- For `DRAFT_EXTEND_V2` replay, use uniform `num_tokens_per_req` semantics for `cu_seqlens_q` and `max_seq_len_q`. The v1 `accept_lens` path is unchanged.
- Make the V2 logits/top-k contract explicit: `EagleDraftWorker` owns the selected-row `softmax + fast_topk` result used by speculative sampling; graph replay should not compute or return full-logits `ret.topk_p` / `ret.topk_index`.
- Replace graph-side full-logits `softmax + topk/reduce` with a cheap max-reduction anchor (`torch.amax`) in the shared draft-extend CUDA graph runner. The anchor still materializes full logits for the CUDA graph dependency/lifetime contract, but avoids producing unused top-k outputs.

## Accuracy Tests

These are accuracy checks; performance claims use the SA-bench results below.

| Setup | Result |
|---|---|
| Qwen3.5 FP8 TP4, NEXTN/spec_v2 d=4, `trtllm_mha`, full GSM8K, `num_examples=1319`, `max_tokens=16384`, `num_threads=500`, `max-running-requests=500` | current score `0.9764079147640792` (`Score: 0.976`, `1314/1314` effective eval items) |
| DeepSeek-R1 FP4 TP8, Spec V2 EAGLE, `trtllm_mla`, full GSM8K, `num_examples=1319`, `max_tokens=16384`, `num_threads=64`, `max-running-requests=64` | current score `0.9665144596651446` vs baseline reference `0.9703196347031964` (`1314/1314` effective eval items; delta `-0.0038`, within single-run noise) |
| DeepSeek-R1 FP4 TP8, Spec V2 EAGLE, `trtllm_mla`, full GPQA, `num_examples=198`, `repeat=8`, `max_tokens=32768`, `num_threads=32`, `max-running-requests=32` | current mean score `0.6868686868686869` vs baseline reference `0.6887626262626264` (delta `-0.0019`, within repeat-to-repeat noise) |

<details>
<summary>Qwen3.5 full GSM8K benchmark log</summary>

```text
GSM8K Config: endpoint=http://localhost:8000; num_examples=1319; max_tokens=16384; num_threads=500; num_shots=5; temperature=default; top_p=default; top_k=default
Running GSM8K evaluation...
/usr/local/lib/python3.12/dist-packages/torchao/quantization/quant_api.py:1731: SyntaxWarning: invalid escape sequence '\.'
  """Configuration class for applying different quantization configs to modules or parameters based on their fully qualified names (FQNs).
Downloading from https://raw.githubusercontent.com/openai/grade-school-math/master/grade_school_math/data/test.jsonl to /tmp/test.jsonl

/tmp/test.jsonl:   0%|          | 0.00/237k [00:00<?, ?B/s]
/tmp/test.jsonl: 732kB [00:00, 33.5MB/s]                   
ChatCompletionSampler initialized with self.system_message=None self.temperature=0.0 self.max_tokens=16384 self.reasoning_effort=None self.extra_body=None

  0%|          | 0/1314 [00:00<?, ?it/s]
  0%|          | 1/1314 [00:43<15:45:39, 43.21s/it]
  0%|          | 2/1314 [00:54<8:57:14, 24.57s/it] 
  0%|          | 3/1314 [01:44<13:05:56, 35.97s/it]
  0%|          | 6/1314 [02:00<5:44:31, 15.80s/it] 
  1%|          | 8/1314 [03:31<9:53:27, 27.26s/it]
  1%|          | 12/1314 [03:37<4:57:59, 13.73s/it]
  7%|▋         | 96/1314 [04:17<25:46,  1.27s/it]  
  9%|▉         | 115/1314 [04:28<22:09,  1.11s/it]
 48%|████▊     | 636/1314 [04:30<01:16,  8.91it/s]
 48%|████▊     | 636/1314 [04:49<01:16,  8.91it/s]
 60%|██████    | 792/1314 [04:51<01:01,  8.43it/s]
100%|██████████| 1314/1314 [04:51<00:00,  4.51it/s]
Total latency: 291.452 s
Score: 0.976
Output throughput: 11812.895 token/s
[METRIC] gsm8k_score=0.9764079147640792 labels={"model": "Qwen/Qwen3.5-397B-A17B-FP8", "eval": "gsm8k"}
[METRIC] gsm8k_latency=291.4521685200016 labels={"model": "Qwen/Qwen3.5-397B-A17B-FP8", "eval": "gsm8k"}
Writing report to /tmp/gsm8k_Qwen_Qwen3.5-397B-A17B-FP8.html
{'score:std': np.float64(0.15177449967020776), 'score': np.float64(0.9764079147640792), 'latency': 291.4521685200016, 'output_throughput': 11812.895465774252}
Writing results to /tmp/gsm8k_Qwen_Qwen3.5-397B-A17B-FP8.json
Results saved to: /logs/accuracy/gsm8k_Qwen_Qwen3.5-397B-A17B-FP8.json
GSM8K evaluation complete
```

</details>

<details>
<summary>DeepSeek-R1 full GSM8K current benchmark log</summary>

```text
GSM8K Config: endpoint=http://localhost:8000; num_examples=1319; max_tokens=16384; num_threads=64; num_shots=5; temperature=default; top_p=default; top_k=default
Running GSM8K evaluation...
/usr/local/lib/python3.12/dist-packages/torchao/quantization/quant_api.py:1731: SyntaxWarning: invalid escape sequence '\.'
  """Configuration class for applying different quantization configs to modules or parameters based on their fully qualified names (FQNs).
Downloading from https://raw.githubusercontent.com/openai/grade-school-math/master/grade_school_math/data/test.jsonl to /tmp/test.jsonl

/tmp/test.jsonl:   0%|          | 0.00/237k [00:00<?, ?B/s]
/tmp/test.jsonl: 732kB [00:00, 33.0MB/s]
ChatCompletionSampler initialized with self.system_message=None self.temperature=0.0 self.max_tokens=16384 self.reasoning_effort=None self.extra_body=None

  0%|          | 0/1314 [00:00<?, ?it/s]
  0%|          | 1/1314 [00:08<3:00:53,  8.27s/it]
  0%|          | 3/1314 [00:10<1:02:27,  2.86s/it]
  1%|          | 8/1314 [00:16<38:16,  1.76s/it]
  3%|▎         | 36/1314 [00:32<15:50,  1.34it/s]
  7%|▋         | 89/1314 [01:04<13:05,  1.56it/s]
  9%|▉         | 115/1314 [01:22<13:14,  1.51it/s]
 29%|██▊       | 376/1314 [01:29<02:08,  7.28it/s]
 32%|███▏      | 421/1314 [01:32<01:50,  8.05it/s]
 32%|███▏      | 421/1314 [01:49<01:50,  8.05it/s]
 48%|████▊     | 637/1314 [02:26<02:09,  5.22it/s]
 63%|██████▎   | 831/1314 [02:42<01:11,  6.72it/s]
 79%|███████▉  | 1044/1314 [02:47<00:26, 10.15it/s]
 89%|████████▉ | 1172/1314 [02:51<00:11, 12.28it/s]
 90%|████████▉ | 1179/1314 [02:59<00:13,  9.84it/s]
 99%|█████████▉| 1299/1314 [03:04<00:01, 11.89it/s]
100%|██████████| 1314/1314 [03:04<00:00,  7.12it/s]
Total latency: 184.700 s
Score: 0.967
Output throughput: 4047.402 token/s
[METRIC] gsm8k_score=0.9665144596651446 labels={"model": "deepseek-ai/DeepSeek-R1", "eval": "gsm8k"}
[METRIC] gsm8k_latency=184.6997149399831 labels={"model": "deepseek-ai/DeepSeek-R1", "eval": "gsm8k"}
Writing report to /tmp/gsm8k_deepseek-ai_DeepSeek-R1.html
{'score:std': np.float64(0.17990069183674134), 'score': np.float64(0.9665144596651446), 'latency': 184.6997149399831, 'output_throughput': 4047.402023565183}
Writing results to /tmp/gsm8k_deepseek-ai_DeepSeek-R1.json
Results saved to: /logs/accuracy/gsm8k_deepseek-ai_DeepSeek-R1.json
GSM8K evaluation complete
```

</details>

## Performance Validation and Profiling

These results validate this PR configuration and the shared V2 graph logits/top-k cleanup. They are not intended as a generic Spec CUDA Graph speedup claim for every backend/model.

Before (two kernels cunn_SoftMaxForward | ReduceOp<float, at::native::MaxOps...> dead top-k work):
image
image
After (cheap max-reduction anchor (`torch.amax`) in the shared draft-extend CUDA graph runner. Materializes full logits for the CUDA graph dependency/lifetime contract, but avoids producing unused top-k outputs):
image

### Qwen3.5 / TRTLLM MHA

SA-bench E2E validation on Qwen3.5 FP8 TP4, NEXTN/spec_v2 d=4, `trtllm_mha`, GB200, `isl=1000`, `osl=1000`:

| Concurrency | Baseline TPOT | Current TPOT | TPOT delta | Current output TPS | Current accept len |
|---:|---:|---:|---:|---:|---:|
| 4 | 4.399 | 4.178 | -5.03% | 812.6 | 2.952 |
| 8 | 4.698 | 4.634 | -1.36% | 1419.5 | 2.890 |
| 16 | 5.756 | 5.607 | -2.58% | 2487.0 | 2.920 |
| 32 | 7.034 | 6.974 | -0.86% | 3906.5 | 2.954 |
| 64 | 9.279 | 9.089 | -2.05% | 5484.6 | 2.989 |
| 96 | 11.629 | 11.206 | -3.64% | 7048.9 | 2.941 |
| 128 | 12.987 | 12.528 | -3.53% | 8059.7 | 2.934 |
| 192 | 16.698 | 16.284 | -2.48% | 9461.7 | 2.946 |
| 256 | 19.801 | 19.218 | -2.94% | 10610.3 | 2.906 |

Targeted c=64 torch profiler data confirms that the PR removes graph-side full-logits `softmax + fast_topk` work and replaces it with a cheaper full-buffer anchor:

| Scope | GPU time / 100 calls | Per call |
|---|---:|---:|
| baseline eager `DRAFT_EXTEND_V2` annotation | `131.780 ms` | `1.318 ms` |
| previous full-topk graph replay | `159.218 ms` | `1.592 ms` |
| current max-reduction-anchored graph replay | `115.073 ms` | `1.151 ms` |

### DeepSeek-R1 / TRTLLM MLA

The max-reduction anchor cleanup is implemented in the shared `DRAFT_EXTEND_V2` graph runner, so it can affect backends beyond `trtllm_mha`. Additional validation covers DeepSeek-R1 FP4 TP8, Spec V2 EAGLE, `trtllm_mla`, GB200, `isl=1024`, `osl=1024`, with `cuda-graph-bs=[32,64,128,256]` and `max-running-requests=256`.

| Concurrency | TPOT delta | Output TPS delta |
|---:|---:|---:|
| 32 | +0.20% | +2.17% |
| 64 | -2.49% | +1.19% |
| 128 | -1.49% | +1.29% |
| 256 | +0.17% | +3.03% |

Output throughput improved at all four tested concurrencies. TPOT was mixed: c=64/c=128 improved, while c=32/c=256 were effectively neutral. MLA c=512 profiling also shows the removed full-logits work is not free on this backend: graph-correlated GPU time decreased from `400.105 ms / 100` replays to `344.707 ms / 100` replays after replacing full-logits `softmax + MaxOps` with the max-reduction anchor.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests). Draft PR: no unit test added yet; a targeted registered test may be needed after maintainers agree on the direction.
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). No user-facing docs updated yet because this is still a draft correctness/workaround PR.
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so. Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`.
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #26914236449](https://github.com/sgl-project/sglang/actions/runs/26914236449)
Latest PR Test (Extra): :x: [Run #26914235320](https://github.com/sgl-project/sglang/actions/runs/26914235320)

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 1]
```diff
@@ -313,6 +314,7 @@ def init_cuda_graphs(self):
         supports_cuda_draft_extend_graph = (_is_cuda or _is_musa) and (
             isinstance(self.draft_extend_attn_backend, TritonAttnBackend)
             or isinstance(self.draft_extend_attn_backend, TRTLLMMLABackend)
+            or isinstance(self.draft_extend_attn_backend, TRTLLMHAAttnBackend)
         )
```
Comment (merrymercy) [inline python/sglang/srt/speculative/eagle_worker_v2.py]:
```suggestion
        supports_cuda_draft_extend_graph = (_is_cuda or _is_musa) and isinstance(
            self.draft_extend_attn_backend,
            (
                TritonAttnBackend,
                TRTLLMMLABackend,
                TRTLLMHAAttnBackend,
            ),
        )
```

group the isinstance call

**Comment by YAMY1234** [comment]
@merrymercy Thanks for the feedback! Updated🙏

**Comment by merrymercy** [comment]
@YAMY1234 @Fridge003 can we try to fix the conflicts and merge this PR sooner?

**Review by Fridge003 — APPROVED** [review]
